rotkehlchen/chain/decoding/decoder.py

```diff
@@ -285,6 +288,52 @@ def _decode_transaction_from_context(
         - `reload_decoders` lists decoder names that should be reloaded (or `None`).
         """
 
+    def _maybe_load_or_purge_events_from_db(
+            self,
+            transaction: T_Transaction,
+            tx_ref: T_TxHash,
+            location: BLOCKCHAIN_LOCATIONS_TYPE,
+            ignore_cache: bool,
+            delete_customized: bool,
+    ) -> list[T_Event] | None:
+        """Load events from the DB for the given tx if they are already decoded and return them,
+        or purge them from the DB if ignore_cache is True.
+        Returns the list of events or None if the tx was not decoded or the events were purged.
+        """
+        with self.database.conn.read_ctx() as cursor:
+            tx_id = transaction.get_or_query_db_id(cursor)
+
+        mappings_table = 'solana_tx_mappings' if location == Location.SOLANA else 'evm_tx_mappings'
+        if ignore_cache is True:  # delete all decoded events
+            with self.database.user_write() as write_cursor:
+                self.dbevents.delete_events_by_tx_hash(
+                    write_cursor=write_cursor,
+                    tx_hashes=[tx_ref],
+                    location=location,
+                    delete_customized=delete_customized,
+                )
+                write_cursor.execute(
+                    f'DELETE from {mappings_table} WHERE tx_id=? AND value IN (?, ?)',
+                    (tx_id, TX_DECODED, TX_SPAM),
+                )
+        else:  # see if events are already decoded and return them
+            with self.database.conn.read_ctx() as cursor:
+                cursor.execute(
+                    f'SELECT COUNT(*) from {mappings_table} WHERE tx_id=? AND value=?',
+                    (tx_id, TX_DECODED),
+                )
+                if cursor.fetchone()[0] != 0:  # already decoded and in the DB
+                    return self.dbevents.get_history_events_internal(  # type: ignore[return-value]
+                        cursor=cursor,
+                        filter_query=SolanaEventFilterQuery.make(
+                            signatures=[tx_ref],  # type: ignore[list-item]
+                        ) if location == Location.SOLANA else EvmEventFilterQuery.make(
+                            tx_hashes=[tx_ref],  # type: ignore[list-item]
+                        ),
```

Comment:
this can be abstracted in a better way as this assumes the chain can only evmlike/solana